### PR TITLE
Provide a fix to the issue of HTML 5 video not playing #473

### PR DESCRIPTION
**Pull request by dapacheco**
Task/Issue URL: #473
Tech Design URL: 
CC: 

**Description**:
By design, media content requires a user _gesture_ for it to be enabled in a WebView. It is unclear what constitutes a _gesture_ from the API docs.

This setting can be overridden by using the following API - [setMediaPlaybackRequiresUserGesture(boolean)](
https://developer.android.com/reference/android/webkit/WebSettings.html#setMediaPlaybackRequiresUserGesture(boolean))

The default for this setting is _true_, which explains the current behaviour in the app and why toggling the sound allows the content to be played.

[This SO answer](https://stackoverflow.com/a/34535582/3861647) highlights this issue -  and provides a link to a bug that has been raised against Google and their decision to maintain this behaviour: https://bugs.chromium.org/p/chromium/issues/detail?id=178297

**Steps to test this PR**:
1. Navigate to a page with HTML 5 video - https://tekeye.uk/html/html5-video-test-page 
1. Try to play the media
1. Launch settings and disable 'Require gesture for media playback'
1. Reload tab
1. Try to play the media - the video should now play

To verify autoplay behaviour for a video use this site - http://camendesign.com/code/video_for_everybody/test.html, when gesture is required the video won't autoplay. However, when the gesture is not required loading the page will autoplay the video.

The following was taken into consideration:
- Setting the `setMediaPlaybackRequiresUserGesture` value to `false` would enable `autoplay` for all media which may have an undesired side-effect. This is why I added a toggle under settings.
- Using a mechanism similar to a theme change to apply the setting to existing tabs would force a reload of the page without alerting the user beforehand. So I went with an implementation that requires the user to reload the page after making the setting change.

There is still an issue of discoverability of this setting and how changing it modifies the behaviour of the web view.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by subsymbolic** [comment]
Thanks for your PR @dapacheco, your investigation into this issue is really helpful. In terms of desirable UX we would like autoplay off by default and pressing play to start a video rather than a setting. If that doesn't look feasible, we'll need to investigate other options in-house.

File: app/src/main/java/com/duckduckgo/app/settings/SettingsActivity.kt [hunk 1]
```diff
@@ -56,12 +56,16 @@ class SettingsActivity : DuckDuckGoActivity(), SettingsAutomaticallyClearWhatFra
...
+        viewModel.onPlayMediaRequireGestureChanged(isChecked)
+    }
+
     private val lightThemeToggleListener = OnCheckedChangeListener { _, isChecked ->
         viewModel.onLightThemeToggled(isChecked)
     }
 
     private val autocompleteToggleListener = OnCheckedChangeListener { _, isChecked ->
-        viewModel.onAutocompleteSettingChanged(isChecked)
+        viewModel.onPlayMediaRequireGestureChanged(isChecked)
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/settings/SettingsActivity.kt]: Revert unintentional change.

**Comment by dapacheco** [comment]
I had only tested this issue on a device running Android 9 (a Pixel XL) and had verified the behaviour described so I didn't try any other devices before looking for a possible fix.

However, I've been thinking that this behaviour seems more like a bug with the WebView than anything else.

The autoplay requiring a gesture seems resonable but the inability to play video without first toggling the sound by default seems broken, even if changing `setMediaPlaybackRequiresUserGesture` _fixes_ it.

I found an older device, running 5.0.1, to test this on and it works as expected without any changes. This was also running an older version of Chrome, but after updating to the latest version of Chrome that too worked.

I've also tested a number of emulators and I've not been able to reproduce the broken behaviour. Even one that matches the OS and Chrome version of the physical device I've tested on. Although emulators don't always show the same issues as devices.

Finally, I've just updated the Pixel to Android Q and this works as expected. As such, I'm going to close this PR given that the root cause appears to be a system bug that affects WebViews (I've been testing on a simple WebView app to isolate any DDG app changes from this problem).

**Comment by subsymbolic** [comment]
Thanks for investigating @dapacheco. Much appreciated.

**Pull request closed without merging**